### DISCUSSION TITLE
Inventory saving and loading for players

### DISCUSSION THREAD

**Discussion by dev-hydrogen** [Minestom/Minestom#1198] [Help] [answered]
Is there an existing implementation of vanilla-like inventory loading and saving, to file or a database, that i could use?

**Comment by TheMode** [comment, marked as answer]
You can use the [Tag API](https://wiki.minestom.net/feature/tags)

```java
// Tags represent how the data should be accessed
Tag<List<ItemStack>> tag = Tag.ItemStack("items").list();

// Tag handlers are basically nbt compound holder, but working with tags and mostly hiding the NBT detail
TagHandler handler = TagHandler.newHandler();
handler.setTag(tag, List.of(ItemStack.of(Material.DIAMOND),ItemStack.of(Material.DIAMOND)));

// Serialize to nbt, that you can save in a database
NBTCompound nbt = handler.asCompound();
System.out.println(nbt);

// Deserialize back from nbt
handler = TagHandler.fromCompound(nbt);
List<ItemStack> items = handler.getTag(tag);
```

Then you may also want to save the inventory name, achievable using a second component tag (`Tag.Component`)

**Reply by dev-hydrogen** [reply]
A) How do i get and then later set a list of itemstacks and their slots in a players inventory? (I cant find such method in PlayerInventory at least)
B) How should i save what slot the itemstacks are in?

**Reply by TheMode** [reply]
A) You can loop through the list and set each individual slot
B) You have 2 options 1) somehow serialize an int array with the index of each item 2) make your list size match the inventory size and include air items
